video/out/wayland_common.c

```diff
@@ -1297,3 +1320,39 @@ bool vo_wayland_config (struct vo *vo, uint32_t flags)
 
     return true;
 }
+
+void vo_wayland_request_frame(struct vo *vo, void *data, vo_wayland_frame_cb cb)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+    wl->frame.data = data;
+    wl->frame.function = cb;
+    MP_DBG(wl, "restart frame callback\n");
+    frame_callback(wl, NULL, 0);
+}
+
+bool vo_wayland_wait_frame(struct vo *vo)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+
+    if (!wl->frame.callback || wl->frame.dropping)
+        return false;
+
+    // If mpv isn't receiving frame callbacks, this usually means that mpv
+    // window is not visible and compositor tells kindly to not draw anything.
+    while (!wl->frame.pending) {
+        int64_t timeout = wl->frame.frame_ms;
+
+        if (vo_wayland_poll(vo, timeout) <= 0)
+            break;
+
+        if ((mp_time_us() - wl->frame.last_us) / 1000 < timeout)
+            break;
+    }
```

Comment:
Why have both frame.frame_ms and frame.last_us? Can you just keep the later, and calculate the relative timeout every loop calculation?